File: 3-2.c
#include "prelude.h"

// priorities go from [1, 52]
#define N 53

static char const example[] = "vJrwpWtwJgWrhcsFMMfFFhFp\n"
			      "jqHRNqRjqzjGDLGLrsFMfFZSrLrFZsSL\n"
			      "PmmdzqPrVvPwwTWBwg\n"
			      "wMqvLMZHhHMvwLHjbvcjnnSBnvTQFn\n"
			      "ttgJtRGJQctTZtZT\n"
			      "CrZsJsPPZsGzwwsLwLmpwMDw";

static int priority(char c) {
	assert(c <= 'z' && c >= 'A' && "Unexpected character!");
	return c >= 'a' && c <= 'z' ? c - 'a' + 1 : c - 'A' + 27;
}

static int solve(char *input) {
	int items[N] = {0};
	int sum = 0;

	int n = 0;
	LINES(line, input) {
		int elf = 1 << n++ % 3;
		for (int i = 0; line[i]; i++)
			items[priority(line[i])] |= elf;

		if (n > 0 && n % 3 == 0) {
			for (int i = 1; i < N; i++) {
				if (items[i] == (1 | 2 | 4)) sum += i;
				items[i] = 0;
			}
		}
	}

	return sum;
}

int main() {
	RUN(solve, strdup(example));
	RUN(solve, read_to_string("3.txt"));
	return 0;
}

File: rs/src/bin/3-2.rs
const EXAMPLE: &str = r"vJrwpWtwJgWrhcsFMMfFFhFp
jqHRNqRjqzjGDLGLrsFMfFZSrLrFZsSL
PmmdzqPrVvPwwTWBwg
wMqvLMZHhHMvwLHjbvcjnnSBnvTQFn
ttgJtRGJQctTZtZT
CrZsJsPPZsGzwwsLwLmpwMDw";

const N: usize = 3;

use std::collections::{HashMap, HashSet};

/// Return the numeric priority of each char
fn priority(c: char) -> i32 {
    match c {
        'a'..='z' => 1 + c as i32 - 'a' as i32,
        'A'..='Z' => 27 + c as i32 - 'A' as i32,
        _ => panic!("Invalid character '{}'", c),
    }
}

fn solve(input: &str) -> i32 {
    let mut priorities: Vec<i32> = Vec::new();
    let mut counts: HashMap<char, usize> = HashMap::new();

    for (i, line) in input.lines().enumerate() {
        let mut chars: HashSet<char> = HashSet::new();
        line.chars().for_each(|c| _ = chars.insert(c));

        chars
            .into_iter()
            .for_each(|c| *counts.entry(c).or_insert(0) += 1);

        if i % N == N - 1 {
            counts
                .iter()
                .find(|(&_, &v)| v == N)
                .map(|(&k, _)| priorities.push(priority(k)));
            counts.clear();
        }
    }

    priorities.iter().sum()
}

fn main() -> std::io::Result<()> {
    println!("Example: {}", solve(EXAMPLE));
    println!("Example: {}", solve(&std::fs::read_to_string("../3.txt")?));
    Ok(())
}
